src/api_proxy/auth/auth_token.cc

```diff
@@ -15,188 +15,42 @@
 
-#include <openssl/hmac.h>
-#include <openssl/pem.h>
-#include <cstring>
-#include <string>
+// This header file is for internal. A public header file should not
+// include it.
+#include "src/api_proxy/auth/grpc_internals.h"
 
 extern "C" {
-#include "grpc/grpc.h"
-#include "grpc/support/alloc.h"
-#include "grpc/support/log.h"
-#include "grpc/support/string_util.h"
-#include "grpc/support/sync.h"
-}
-
-#include "grpc_internals.h"
-
-using std::string;
+  #include "grpc/support/alloc.h"
+};
 
 namespace espv2 {
 namespace api_proxy {
 namespace auth {
 
 namespace {
-
-#define GRPC_AUTH_JSON_TYPE_INVALID "invalid"
-
+// Token should expire in 1 hour.
 const gpr_timespec TOKEN_LIFETIME = {3600, 0, GPR_TIMESPAN};
-
-char *GenerateTokenSymmetricKey(const grpc_json *json, const char *audience);
-
-char *GenerateJoseHeader(const char *algorithm);
-char *GenerateJwtClaim(const char *issuer, const char *subject,
-                       const char *audience, gpr_timespec token_lifetime);
-char *GenerateSignatueHs256(const char *data, const char *key);
-string DotConcat(const string &str1, const string &str2);
-
 }  // namespace
 
-// TODO: this function can return a string instead of char* that need to be
-// freed.
-char *get_auth_token(const char *json_secret, const char *audience) {
-  char *scratchpad = gpr_strdup(json_secret);
-  grpc_json *json = grpc_json_parse_string(scratchpad);
-  char *token = nullptr;
-
-  if (GetStringValue(json, "client_secret") != nullptr) {  // Symmetric key.
-    token = GenerateTokenSymmetricKey(json, audience);
-  } else {
-    grpc_auth_json_key json_key =
-        grpc_auth_json_key_create_from_string(json_secret);
-    if (strcmp(json_key.type, GRPC_AUTH_JSON_TYPE_INVALID) != 0) {
-      token = grpc_jwt_encode_and_sign(&json_key, audience, TOKEN_LIFETIME,
-                                       nullptr);
-      grpc_auth_json_key_destruct(&json_key);
-    }
-  }
-
-  if (json != nullptr) grpc_json_destroy(json);
-  gpr_free(scratchpad);
-  return token;
-}
-
-void grpc_free(char *token) { gpr_free(token); }
-
-namespace {
-
-char *GenerateTokenSymmetricKey(const grpc_json *json, const char *audience) {
-  const char *issuer = GetStringValue(json, "issuer");
-  if (issuer == nullptr) {
-    // TODO: move to common logging.
-    gpr_log(GPR_ERROR, "Missing issuer");
-    return nullptr;
-  }
-  const char *subject = GetStringValue(json, "subject");
-  if (subject == nullptr) {
-    gpr_log(GPR_ERROR, "Missing subject");
-    return nullptr;
-  }
+absl::optional<std::string> get_auth_token(const std::string& json_secret,
+                                           const std::string& audience) {
+  grpc_auth_json_key json_key =
+      grpc_auth_json_key_create_from_string(json_secret.c_str());
 
-  char *header = GenerateJoseHeader("HS256");
-  if (header == nullptr) {
-    gpr_log(GPR_ERROR, "Unable to create JOSE header");
-    return nullptr;
+  if (grpc_auth_json_key_is_valid(&json_key) == 0) {
```

Comment:
Are they equivalent? Is json_key.type checked?
- if (strcmp(json_key.type, GRPC_AUTH_JSON_TYPE_INVALID) != 0) { 
- if (grpc_auth_json_key_is_valid(&json_key) == 0) {

Reply:
Yup: https://github.com/grpc/grpc/blob/dd8259f0eefcfc0743646bdc278a16b2cfad8b17/src/core/lib/security/credentials/jwt/json_token.cc#L66